types/convert.go

```diff
@@ -443,12 +443,15 @@ func floatStrToIntStr(sc *stmtctx.StatementContext, validFloat string, oriStr st
 	if err != nil {
 		return validFloat, errors.Trace(err)
 	}
-	if exp > 0 && int64(intCnt) > (math.MaxInt64-int64(exp)) {
-		// (exp + incCnt) overflows MaxInt64.
+	intCnt += exp
+	// intCnt will < 0 when overflow
+	if intCnt < 0 || intCnt > 20 {
+		// MaxInt64's has 19 decimal digit.
+		// And the intCnt may contain the len of `+/-`,
+		// so I use 20 here as the early detection.
 		sc.AppendWarning(ErrOverflow.GenWithStackByArgs("BIGINT", oriStr))
 		return validFloat[:eIdx], nil
 	}
-	intCnt += exp
 	if intCnt <= 0 {
```

Comment:
The `intCnt` will never be negative according to the L448.
Prefer to:
```go
if intCnt > 0 {
	extraZeroCount := ...
} else {
	...
}
```

Reply:
intCnt>=0, exp>=0, however, if exp=I64::MAX, intCnt>0, then their sum will overflow and get negative.

Reply:
intCnt>=0, exp>=0, however, if exp=I64::MAX, intCnt>0, then their sum will overflow and get negative.

Reply:
> intCnt>=0, exp>=0, however, if exp=I64::MAX, intCnt>0, then their sum will overflow and get negative.

If `intCnt < 0` will early return in L451.